Add ROW constructor

lacbs opened the pull request:
It doesn't explicitly set the fields names.
Now it's possible to do nice things like `select row(1, row(9, a, array[], null), row(1, 2)), row(date '2016-04-15') from (values ('a'), ('b')) t(a);`

Notice that the type for ROW will have changes.
Now it's syntax is `ROW<type1, type2...>(field1, field2...)`, but it should be `ROW(field1 type1, field2 type2...)`
Because of the presence of `<` and `>` in the type name, this commit introduces a hack because the typeName is used as an part of variables name in the generated bytecode. Once it's changed to the new syntax, we can remove the hack.

Also, ROW CAST is not completly implemented:
- CAST from one row to another doesn't rename fields.
- It doesn't CAST the arguments types to super types.
- It's not possible to cast from NULL (implicitly) because it generates a type with fields with uppercase letters.
- It's not possible to explicitly cast because the the file `SqlBase.g4` doesn't define the syntax for ROW type.
- Because of the same reason you can get a weird error if you try `select * from (values (row(1,2), 3), (null, 3));`. Since arilift doesn't have the syntax for `ROW<types>(fields)` it throws `ParsingException`.

lacbs commented:
Will solve the conflicts once it's accepted.

Please let me know if/where I need to add more tests. I couldn't find a place to test with `from (values...` clause.
If I try to add to `AbstractTestQueries` and run `TestLocalQueries`, it get `Function "ROW" not found`.

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:
```diff
@@ -488,6 +488,9 @@ public Signature resolveFunction(QualifiedName name, List<TypeSignature> paramet
             String typeName = name.getSuffix().substring(MAGIC_LITERAL_FUNCTION_PREFIX.length());
+            // The RowType have <> that are not allowed in java variable names, so we replace to {}. We can remove this
+            // hack when the type ROW doesn't have them anymore (be ROW(field type+) instead of ROW<type+>(field+))
+            typeName = typeName.replace('{', '<').replace('}', '>');
```
Why do we need this? It shouldn't be necessary, I don't think. In the bytecode generation you should be able to strip the "<" and ">"

lacbs replied:
Unfortunately it's not :(
When it tries to store a variable of type `ROW<..>(...)` (and use the type as part of the name), it fails here: [InvokeInstruction.java](https://github.com/prestodb/presto/blob/c7bf63a2138c2e61f0edb307cefbf6b35fba25c2/presto-bytecode/src/main/java/com/facebook/presto/bytecode/instruction/InvokeInstruction.java#L431)

If I remove that check, we get another error when trying to execute the bytecode.

And if we simply strip, it's going to be complicated to recover the original name.

Fortunately, https://github.com/prestodb/presto/pull/5043 will solve this in part by changing the syntax of row type, and then I can remove that. But, still anything with `.;[/<>` will fail to compile the bytecode.

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:
```diff
@@ -601,6 +604,9 @@ public ScalarFunctionImplementation getScalarFunctionImplementation(Signature si
             String typeName = signature.getName().substring(MAGIC_LITERAL_FUNCTION_PREFIX.length());
+            // The RowType have <> that are not allowed in java variable names, so we replace to {}. We can remove this
+            // hack when the type ROW doesn't have them anymore (be ROW(field type+) instead of ROW<type+>(field+))
+            typeName = typeName.replace('{', '<').replace('}', '>');
```
same

lacbs replied:
ditto

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:
```diff
@@ -739,7 +745,10 @@ public static Signature getMagicLiteralFunctionSignature(Type type)
+        // The RowType have <> that are not allowed in java variable names, so we replace to {}. We can remove this
+        // hack when the type ROW doesn't have them anymore (be ROW(field type+) instead of ROW<type+>(field+))
+        String typeName = type.getTypeSignature().toString().replace('<', '{').replace('>', '}');
+        return new Signature(MAGIC_LITERAL_FUNCTION_PREFIX + typeName,
```
same comment

lacbs replied:
ditto.
I can wait for https://github.com/prestodb/presto/pull/5043 to remove this and then land.

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java:
```diff
@@ -973,7 +975,26 @@ protected Object visitArrayConstructor(ArrayConstructor node, Object context)
+                for (int i = 0; i < n; ++i) {
+                    writeNativeValue(parameterTypes.get(i), blockBuilder, values.get(i));
+                }
+                return blockBuilder;
```
return `blockBuilder.build()`

lacbs replied:
Thanks, forgot it, will do :)

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java:
```diff
@@ -973,7 +975,26 @@ protected Object visitArrayConstructor(ArrayConstructor node, Object context)
+            List<Type> parameterTypes = rowType.getTypeParameters();
+            List<Expression> arguments = node.getItems();
+
+            int n = arguments.size();
```
We try to avoid single letter variable names, so either inline this or name it something more descriptive like "cardinality"

lacbs replied:
will use `cardinality` :)

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/relational/SqlToRowExpressionTranslator.java:
```diff
@@ -572,5 +574,18 @@ protected RowExpression visitArrayConstructor(ArrayConstructor node, Void contex
+                    .map(value -> process(value, context))
+                    .collect(toImmutableList());
+            Type returnType = types.get(node);
+            List<Type> itemsTypes = node.getItems().stream()
```
name this `argumentTypes` instead of `itemsTypes`

lacbs replied:
ok will change, only use itemTypes because `getItems` :P

cberner commented:
Add tests in `TestRowOperators`

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/gen/RowConstructorCodeGenerator.java:
```diff
@@ -0,0 +1,128 @@
+        implements BytecodeGenerator
+{
+    @Override
+    public BytecodeNode generateExpression(Signature signature, BytecodeGeneratorContext context, Type rowType, List<RowExpression> arguments)
```
Instead of using `BytecodeBlock` directly for all of this, take a look at `com.facebook.presto.operator.scalar.ArrayConstructor.generateArrayConstructor()`. It will be much simpler if you use bytecode expressions like that uses

lacbs replied:
Great, it makes total sense to have the calls to to the right function in of BytecodeBlock in one place `SqlTypeBytecodeExpression`. Thanks will use it :D

cberner commented on presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java:
```diff
@@ -956,6 +956,32 @@ public void testArrayConstructor()
+        // the following tests should how when the CAST to ROW is working
+
+        // these two tests don't work because one type is varchar(17) and the other varchar
+        // optimize("ARRAY [ROW(unbound_string, unbound_double), ROW(bound_string, 0.0)]");
```
I think you can fix this by using `CAST(bound_string AS VARCHAR)`

lacbs replied:
that's true, but the problem would still be there, fortunately after https://github.com/prestodb/presto/pull/5043 it should work

cberner commented on presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java:
```diff
@@ -956,6 +956,32 @@ public void testArrayConstructor()
+        // optimize("ARRAY [ROW(bound_string, 0.0), ROW(unbound_string, unbound_double)]");
+
+        // these two tests don't work because the CAST from NULL generate a row with uppercase
+        // field names (row<varchar,double>('FIELD0','FIELD1')), while the row expression
```
I'm a little confused about this comment. `row()` should return a row type with no field names, so where do the "field0", "field1" come from?

lacbs replied:
Even if you don't pass any name, it's populating them :/
See [TypeSignature.java] (https://github.com/prestodb/presto/blob/c7bf63a2138c2e61f0edb307cefbf6b35fba25c2/presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java#L240)

That's why we also have some problems with fields names and lower/upper case.

cberner commented:
Nice work! Had some comments. Let me know when it's ready for review again.

lacbs commented:
Will add tests in `TestRowOperators`. Thanks a lot for the review :D
Will make the changes later this week, probably Friday.
Sorry for my silence here for such a long time.

geraint0923 commented:
@lacbs #5043 has been merged into the master, you can check it out and rebase. Thanks a lot!

cberner commented:
Sorry, didn't see that you had updated this. Github doesn't send any emails when you push commits, so you have to post a comment when you push new ones

lacbs commented:
Oh, I didn't know that, will post a comment next time. Thanks.

cberner commented on presto-main/src/test/java/com/facebook/presto/type/TestRowOperators.java:
```diff
@@ -104,6 +119,9 @@ public void testRowCast()
         assertFunction("cast(test_row(2, cast(null as double)) as row(aa bigint, bb double)).bb", DOUBLE, null);
         assertFunction("cast(test_row(2, 'test_str') as row(aa bigint, bb varchar)).bb", VARCHAR, "test_str");
 
+        // Should fail for ambiguity
```
Add "TODO:" at the beginning of the comment

cberner replied:
Actually file a Github issue instead, and remove this

lacbs replied:
I think this one is ok, I will just remove, since the column names in a table are not case sensitive either.
Or better, I can make it expect an exception like with the hyperloglog test here.

cberner commented on presto-main/src/test/java/com/facebook/presto/type/TestRowOperators.java:
```diff
@@ -169,5 +187,35 @@ public void testRowEquality()
         assertFunction("test_row(TRUE, ARRAY [1]) != test_row(TRUE, ARRAY [1,2])", BOOLEAN, true);
         assertFunction("test_row(1.0, ARRAY [1,2,3], test_row(2,2.0)) != test_row(1.0, ARRAY [1,2,3], test_row(1,2.0))", BOOLEAN, true);
+
+        // Cannot be applied with different field names
```
File a Github issue that coercions between row types need to be implemented, and then remove these commented out tests

cberner commented on presto-main/src/test/java/com/facebook/presto/type/TestRowOperators.java:
```diff
@@ -169,5 +187,35 @@ public void testRowEquality()
+    public void testRowSelect()
+            throws Exception
+    {
+        testQuery("SELECT ROW(1 + 2, CONCAT('a', 'b'))", "SELECT ROW(3, 'ab')");
```
Add these tests to `AbstractTestQueries` instead of here. Then you can remove the `queryRunner`

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/relational/Signatures.java:
```diff
@@ -155,6 +156,11 @@ public static Signature inSignature()
 
+    public static Signature rowConstructorSignature(Type returnType, List<Type> argumentTypes)
+    {
+        return internalScalarFunction(ROW_CONSTRUCTOR, returnType.getTypeSignature(), Lists.transform(argumentTypes, Type::getTypeSignature));
```
We're mostly moving away from `Lists.transform`, so use `argumentTypes.stream().map(Type::getTypeSignature).collect(toImmutableList())`

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java:
```diff
@@ -991,7 +993,26 @@ protected Object visitArrayConstructor(ArrayConstructor node, Object context)
+            int cardinality = arguments.size();
+            List<Object> values = new ArrayList<>(cardinality);
+            for (Expression argument : arguments) {
+                Object value = process(argument, context);
```
You can inline the `value` variable

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java:
```diff
@@ -991,7 +993,26 @@ protected Object visitArrayConstructor(ArrayConstructor node, Object context)
+                Object value = process(argument, context);
+                values.add(value);
+            }
+            if (optimize && hasUnresolvedValue(values)) {
```
Why is this `optimize &&`? I think this check should just be `hasUnresolvedValue(values)`

lacbs replied:
Thanks for catching this. This would fail. I made this function based on visitFunctionCall, them I forgot to remove the `optimize &&`, sorry.

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/gen/RowConstructorCodeGenerator.java:
```diff
@@ -0,0 +1,78 @@
+        BytecodeBlock block = new BytecodeBlock().setDescription("Constructor for " + rowType.toString());
+        CallSiteBinder binder = context.getCallSiteBinder();
+        Scope scope = context.getScope();
+        List<Type> typeParameters = rowType.getTypeParameters();
```
I would name this `types` instead. "type parameter" has several other meanings in the code base, like "T" in "array<T>"

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/gen/RowConstructorCodeGenerator.java:
```diff
@@ -0,0 +1,78 @@
+        Variable blockBuilderVar = scope.createTempVariable(BlockBuilder.class);
+        Binding typesBinding = binder.bind(typeParameters, List.class);
+        block.append(blockBuilderVar.set(
+                newInstance(InterleavedBlockBuilder.class, loadConstant(typesBinding), newInstance(BlockBuilderStatus.class), constantInt(cardinality))));
```
The last argument should be `1` rather than `cardinality`. That argument is the number of complete rows that are expected to be written to the block

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/gen/RowConstructorCodeGenerator.java:
```diff
@@ -0,0 +1,78 @@
+                newInstance(InterleavedBlockBuilder.class, loadConstant(typesBinding), newInstance(BlockBuilderStatus.class), constantInt(cardinality))));
+
+        for (int i = 0; i < cardinality; ++i) {
+            Type elementType = typeParameters.get(i);
```
I would name this `fieldType` instead, since the parts of a row are referred to as "fields"

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/gen/RowConstructorCodeGenerator.java:
```diff
@@ -0,0 +1,78 @@
+                block.append(blockBuilderVar.invoke("appendNull", BlockBuilder.class).pop());
+            }
+            else {
+                Variable elementVar = scope.createTempVariable(javaType);
```
We try to avoid abbreviations in the code base, so I would call this `fieldValue` instead of `elementVar`

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/gen/RowConstructorCodeGenerator.java:
```diff
@@ -0,0 +1,78 @@
+        int cardinality = arguments.size();
+
+        block.comment("BlockBuilder blockBuilder = new InterleavedBlockBuilder(types, new BlockBuilderStatus(), n);");
+        Variable blockBuilderVar = scope.createTempVariable(BlockBuilder.class);
```
We try to avoid abbreviations, so I would just call this `blockBuilder`

cberner commented:
Mainly stylistic comments. Let me know when you've addressed them and I'll take another look

lacbs commented:
Will address the comments, thanks for the quick review :)

lacbs commented on presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java:
```diff
@@ -1136,6 +1136,37 @@ public void testArrayConstructor()
+        // generate a with with lowercase field names (row(field0 varchar, field1 double))
+        // Notice that parseRowTypeSignature calls createNamedTypeParameters in TypeSignature
+        // optimize("ARRAY [NULL, ROW(unbound_string, unbound_double)]");
+        // optimize("ARRAY [ROW(unbound_string, unbound_double), NULL]");
```
Should I create an issue for this too?

cberner replied:
Ya, generally I don't like having commented out tests because we tend to forget about them. With an issue we're more likely to remember, and can track it

lacbs commented:
Just pushed the version addressing the comments.
Let me know if you want me to create an issue for auto casting row types.

Also, instead of creating an issue, I think having field case insensitive is expected, so I added this piece of code:
```
        try {
            assertFunction("CAST(ROW(1, 2) AS ROW(a BIGINT, A DOUBLE)).a", BIGINT, 1L);
            fail("fields in Row are case insensitive"); // This throws an Error, not an Exception
        }
        catch (Exception e) {
            // Expected
        }
```

lacbs commented:
@cberner, I actually run into a issue, `testRows` (from `TestLocalQueries`) outputs
```
org.skife.jdbi.v2.exceptions.UnableToCreateStatementException: org.h2.jdbc.JdbcSQLException: Function "ROW" not found; SQL statement:
SELECT ROW(3, 'ab') [90022-189] [statement:"SELECT ROW(3, 'ab')", located:"SELECT ROW(3, 'ab')", rewritten:"SELECT ROW(3, 'ab')", arguments:{ positional:{}, named:{}, finder:[]}]
```
I think it's because you can evaluate `row(x)` as `function_name(x)`. Do you know how I get around this?

Starting a server and connecting to it works fine.

lacbs commented:
Got the problem, the second argument of assertQuery uses H2. I'm using `json_format(cast(_ as json))` in the first argument to match the expected result.
All tests should pass now.

cberner commented:
Merged. Thanks!

lacbs commented:
I just created the issue https://github.com/prestodb/presto/issues/5233 and was going to remove the tests, but this is already merged... :P
Thank you for the review.

cberner commented:
Yep, I removed them for you while I was merging it :)

lacbs commented:
Thanks =)
